### ISSUE TITLE
The testing method WithCustomState doesn't work properly when the state is a string resulting from a concatenation.

### ISSUE THREAD

**Issue by NerdBrez** [FluentValidation/FluentValidation#1658]
How to reproduce: 
Add a validation that has a custom state as follow:
` .WithState(x => "Test" + "123");`

Add a unit test for this validation and try to assert it as follow:
`.WithCustomState("Test123");`

The test will fail with the message:
```
Message: 
        FluentValidation.TestHelper.ValidationTestException : Expected custom state of 'Test123'. Actual state was 'Test123'
```

I have got the same error using no matter what form of concatenation: StrigBuilder, String.Format, $ operator, + operator.

If you remove the concatenation from the validator and use it just as follow, the test works fine:
` .WithState(x => "Test123");`

The nuget version I used is the latest at this moment: `9.5.1`.

Could someone check that please? 
Thanks.

**Comment by JeremySkinner** [comment]
Hello,

I tried this and it worked correctly with a concatenated string:

```csharp
[Fact]
public void Test_custom_state_with_concatenated_string() {
	var validator = new InlineValidator<Person>();
	validator.RuleFor(x => x.Surname).NotNull().WithState(x => "Test" + "123");
	var result = validator.TestValidate(new Person());
	result.ShouldHaveValidationErrorFor(x => x.Surname).WithCustomState("Test123");
}
``` 

The test passed correctly without throwing an exception. Please provide accurate steps to reproduce the issue. Thanks

**Comment by NerdBrez** [comment]
Hello, I'm sorry, I made a mistake in my first post, in fact to reproduce the problem the validator must concatenate a string with an integer value as follow:
`.WithState(x => "Test" + 123.ToString());`
or
`.WithState(x => $"Test{123}");`

Thanks!

**Comment by JeremySkinner** [comment]
I see what's happening. Internally a reference equality check is being performed. This example illustrates the problem:

```csharp
object s1 = "Test" + 123.ToString();
object s2 = "Test123";

bool check1 = s1 == s2; // False
bool check2 = s2.Equals(s1); // True
```
Becuase the `==` operator is used against variables of type `object`, the wrong type equality check is performed. Internally this is what FluentValidation is doing. 

When explicitly typed as strings, the result is as expected:

```csharp
string s1 = "Test" + 123.ToString();
string s2 = "Test123";

bool check1 = s1 == s2; // True
```

Internally FluentValidation needs to be changed to use `.Equals`, and also should optionally accept an `IEqualityComparer`. To work around this for now, you can add your own test extension that takes a comparer:

```csharp
public static class MyExtensions {
  public static IEnumerable<ValidationFailure> WithCustomState(this IEnumerable<ValidationFailure> failures, object expectedCustomState, IEqualityComparer comparer) {
	  return failures.When(failure => comparer.Equals(failure.CustomState, expectedCustomState), $"Expected custom state of '{expectedCustomState}'. Actual state was '{{State}}'");
  }
}
```

...which you can then use like this:

```csharp
result.ShouldHaveValidationErrorFor(x => x.Surname).WithCustomState("Test123", StringComparer.Ordinal);
```

**Comment by NerdBrez** [comment]
I see. Thank you very much!

**Comment by JeremySkinner** [comment]
Implemented in 9.5.2
